sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/WriteToDataSourceV2Exec.scala

```diff
@@ -41,18 +46,114 @@ case class WriteToDataSourceV2(batchWrite: BatchWrite, query: LogicalPlan)
   override def output: Seq[Attribute] = Nil
 }
 
+case class AppendDataExec(
+    table: SupportsBatchWrite,
+    writeOptions: DataSourceOptions,
+    query: SparkPlan) extends V2TableWriteExec with BatchWriteHelper {
+
+  override protected def doExecute(): RDD[InternalRow] = {
+    val batchWrite = newWriteBuilder() match {
+      case builder: SupportsSaveMode => // TODO: Remove this
+        builder.mode(SaveMode.Append).buildForBatch()
+
+      case builder =>
+        builder.buildForBatch()
+    }
+    doWrite(batchWrite)
+  }
+}
+
+case class OverwriteByExpressionExec(
+    table: SupportsBatchWrite,
+    filters: Array[Filter],
+    writeOptions: DataSourceOptions,
+    query: SparkPlan) extends V2TableWriteExec with BatchWriteHelper {
+
+  private def isTruncate(filters: Array[Filter]): Boolean = {
+    filters.length == 1 && filters(0).isInstanceOf[AlwaysTrue]
+  }
+
+  override protected def doExecute(): RDD[InternalRow] = {
+    val batchWrite = newWriteBuilder() match {
+      case builder: SupportsTruncate if isTruncate(filters) =>
+        builder.truncate().buildForBatch()
+
+      case builder: SupportsOverwrite =>
+        builder.overwrite(filters).buildForBatch()
+
+      case builder: SupportsSaveMode => // TODO: Remove this
+        builder.mode(SaveMode.Overwrite).buildForBatch()
+
+      case _ =>
+        throw new SparkException(s"Table does not support dynamic partition overwrite: $table")
```

Comment:
we should fail earlier, at analysis time, or at least planning time.

Reply:
Can we do this as a follow-up? I think this check should use capabilities and I don't want to add that API in this commit. Capabilities will be my next PR.

Reply:
sure, SGTM

Reply:
Opened [SPARK-26811](https://issues.apache.org/jira/browse/SPARK-26811) to track earlier analysis exceptions.